flink-contrib/flink-timerserivce-rocksdb/src/main/java/org/apache/flink/contrib/streaming/api/operators/RocksDBInternalTimerService.java

```diff
@@ -0,0 +1,797 @@
+import org.apache.flink.core.memory.DataOutputViewStreamWrapper;
+import org.apache.flink.runtime.state.KeyGroupRange;
+import org.apache.flink.runtime.state.KeyGroupRangeAssignment;
+import org.apache.flink.streaming.api.operators.InternalTimer;
+import org.apache.flink.streaming.api.operators.InternalTimerService;
+import org.apache.flink.streaming.api.operators.KeyContext;
+import org.apache.flink.streaming.runtime.tasks.ProcessingTimeService;
+import org.apache.flink.util.Preconditions;
+import org.rocksdb.ColumnFamilyDescriptor;
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.ColumnFamilyOptions;
+import org.rocksdb.CompactionStyle;
+import org.rocksdb.DBOptions;
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksDBException;
+import org.rocksdb.RocksIterator;
+import org.rocksdb.StringAppendOperator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * {@link InternalTimerService} that stores timers in RocksDB.
+ */
+public class RocksDBInternalTimerService<K, N> extends InternalTimerService<K, N> {
+	
+	private static Logger LOG = LoggerFactory.getLogger(RocksDBInternalTimerService.class);
+	
+	/** The data base where stores all timers */
+	private final RocksDB db;
+	
+	/** The path where the rocksdb locates */
+	private final Path dbPath;
+
+	/**
+	 * The in-memory heaps backed by rocksdb to retrieve the next timer to trigger. Each
+	 * partition's leader is stored in the heap. When the timers in a partition is changed, we
+	 * will change the partition's leader and update the heap accordingly.
+	 */
+	private final int numPartitions;
+	private final PersistentTimerHeap eventTimeHeap;
+	private final PersistentTimerHeap processingTimeHeap;
+	
+	private static int MAX_PARTITIONS = (1 << 16);
+
+	public RocksDBInternalTimerService(
+			int totalKeyGroups,
+			KeyGroupRange keyGroupRange,
+			KeyContext keyContext,
+			ProcessingTimeService processingTimeService,
+			Path dbPath) {
+
+		super(totalKeyGroups, keyGroupRange, keyContext, processingTimeService);
+		
+		this.dbPath = dbPath;
+		
+		try {
+			FileSystem fileSystem = this.dbPath.getFileSystem();
+			if (fileSystem.exists(this.dbPath)) {
+				fileSystem.delete(this.dbPath, true);
+			}
+			
+			fileSystem.mkdirs(dbPath);
+		} catch (IOException e) {
+			throw new RuntimeException("Error while creating directory for rocksdb timer service.", e);
+		}
+
+		ColumnFamilyOptions columnFamilyOptions = new ColumnFamilyOptions()
+				.setMergeOperator(new StringAppendOperator())
+				.setCompactionStyle(CompactionStyle.UNIVERSAL);
+		ColumnFamilyDescriptor defaultColumnDescriptor = new ColumnFamilyDescriptor("default".getBytes(), columnFamilyOptions);
+
+		DBOptions dbOptions = new DBOptions()
```

Comment:
I wonder if we should keep a reference to this so that we can dispose the native object on close? Again, this problem goes away when reusing the RocksDB instance.